### ISSUE TITLE
d3dcompiler_43-47 broken

### ISSUE THREAD

**Issue by RobertBColton** [msys2/MINGW-packages#5330]
It seems that after my latest `pacman -Syu` system upgrade, not all is well. The D3D shader compiler seems to have been obscurely broken. The D3D11 dependencies are all fine and well, but I am suddenly getting `E_INVALIDARG` returned from `D3DReflect` in an app that has been working fine for two years.

When I first got the issue, there was no change to my installed drivers or anything, simply an MSYS2 upgrade. Since then I have tried upgrading the graphics card drivers, but it is still broken.

This is all of the dependencies from MinGW64 that I link:
`override LDLIBS += -ldxgi -ldxguid -ld3d11 -lD3DCompiler`

It seems `-lD3DCompiler` is actually matching `libd3dcompiler_47.a`.

This is on Windows 10 64 bit, and I have an AMD Radeon 290 series graphics card. Direct3D9, OpenGL compatibility and core contexts all work fine and dandy. I seem to also have no problems with other Direct3D11 applications or games, only this one using the D3D shader compiler provided by MinGW/MSYS.

**Comment by RobertBColton** [comment]
Ok, I actually managed to find out the issue. The D3D compiler version 42 and lower does not have the above and my game is working again, no problems with `D3DReflect(...)`. Anything later than 42-47 and my game goes back to having the above issue again. I assume this to be some MinGW/MSYS2 change that broke the binaries.

**Comment by RobertBColton** [comment]
I believe I tracked it down to https://github.com/mirror/mingw-w64/commit/2d4e517ad0c7a9f0bd7001c42e6c131b977c15d9#diff-ee7640f2e7527e2f7b11c66688d8a3ff which ended up in MSYS2.

**Comment by RobertBColton** [comment]
We finally tracked down the real issue: https://github.com/enigma-dev/enigma-dev/pull/1736

The GUID for the shader reflection interface is outdated in `dxguid.lib` provided by MinGW.

**Comment by lazka** [comment]
Was there any communication with mingw-w64 upstream about this? Should this stay open?

**Comment by Biswa96** [comment]
Just ran the `make` command in latest enigma-dev repository and it does not show any error.

**Comment by lazka** [comment]
thanks. Let's assume this is fixed then. If not please try to file this with mingw-w64 as they might know more.

**Comment by RobertBColton** [comment]
Yeah https://github.com/enigma-dev/enigma-dev/pull/1736 patched around it, that's why make would work now, but I don't know if it's still an actual mingw issue. What repo are you saying to file this at?

**Comment by Hexlord** [comment]
-DD3D_COMPILER_VERSION=47 concluded my day of debugging, thank you sir @RobertBColton 
A shame this stuff happens on msys2 + clang :( I wonder why it never happened on msvc

**Comment by Biswa96** [comment]
`D3D_COMPILER_VERSION` is defined in [d3dcompiler.h][1] file. I have compiled enigma-dev successfully with reverting the commit that @RobertBColton mentioned in previous comment. Would you like to provide a simple minimal code that reproduce your issue? Also make sure you are using all the latest packages.

[1]: https://sourceforge.net/p/mingw-w64/mingw-w64/ci/master/tree/mingw-w64-headers/include/d3dcompiler.h

**Comment by Hexlord** [comment]
@Biswa96 you are right, it is defined in the header, I wonder why me adding add_compile_definitions(D3D_COMPILER_VERSION=47) to my root cmakelists fixed the issue, I guess it is defaulted somewhere before the d3dcompiler.h to some other value (but I was unable to find it with text search, bgfx d3dcompiler.h also has unconditional define of 47 version (unlike mingw, respecting already existing define) fyi @bkaradzic).

Sorry, I never used ENIGMA engine, the problem is here: https://github.com/bkaradzic/bgfx/blob/master/tools/shaderc/shaderc_hlsl.cpp#L415 this loop was infinite, as the target function wrote to the d3d struct which had #ifdef locking out one of the fields (the last one), thus being linked against up-to-date d3dcompiler led to it overwritting loop index ii with zero each time, crazy stuff. The interesting part is I could not reproduce it in debug mode (probably due to debug probes messing with stack layout, so it was overwriting probes instead), and had to attach to process to finally see the problem - reset of the loop index. The watch of ii being changed actually happened on some SSE asm instructions (some unpacking, if I recall), so maybe the issue is somehow connected to clang SSE optimizations.

Reproduce is probably: clone bgfx, build it using msys2 mingw 64, try using shaderc to compile dx11, put --verbose and observe infinite loop. 

**tl;dr**: 
D3D11_SIGNATURE_PARAMETER_DESC 'last field is ifdef-ed, which makes bad things happen when mismatching link and compile versions.
-DD3D_COMPILER_VERSION=47 is the way to go, I have no temptation of digging any deeper into the roots of the issue :D

**Comment by Biswa96** [comment]
I can't compile bgfx. I followed [this][1]:

```sh
git clone --depth=1 git://github.com/bkaradzic/bx.git
git clone --depth=1 git://github.com/bkaradzic/bimg.git
git clone --depth=1 git://github.com/bkaradzic/bgfx.git
cd bgfx
make
make mingw-gcc-release64
```

And go this error:

```sh
/bin/sh: /bin/x86_64-w64-mingw32-g++: No such file or directory
```

[1]: https://bkaradzic.github.io/bgfx/build.html

**Comment by Biswa96** [comment]
I can't compile bgfx. I followed [this][1]:

```sh
git clone --depth=1 git://github.com/bkaradzic/bx.git
git clone --depth=1 git://github.com/bkaradzic/bimg.git
git clone --depth=1 git://github.com/bkaradzic/bgfx.git
cd bgfx
make
make mingw-gcc-release64
```

And go this error:

```sh
/bin/sh: /bin/x86_64-w64-mingw32-g++: No such file or directory
```

BTW, I understand the issue that adding `D3D_COMPILER_VERSION == 47` fixes it. The headers are imported from wine. So, I would discuss with them first.

[1]: https://bkaradzic.github.io/bgfx/build.html

**Comment by Hexlord** [comment]
I build bgfx using cmake + msys2 + mingw64 + manually built llvm inside CLion
 MSYS  `pacman -Syuu` -> `pacman -Syuu` 
        -> `pacman -S --needed base-devel mingw-w64-i686-toolchain mingw-w64-x86_64-toolchain git mingw-w64-x86_64-pkg-config mingw-w64-x86_64-protobuf mingw-w64-x86_64-openssl mingw-w64-x86_64-icu mingw-w64-x86_64-gcc mingw-w64-i686-cmake mingw-w64-x86_64-cmake autoconf automake libtool make unzip pacman -S git mercurial cvs wget p7zip perl ruby`
        -> uncomment #MSYS2_PATH_TYPE=inherit for mingw64.ini and mingw32.ini (but NOT msys2)
    LLVM with extras built with GCC (-GNinja -Wa,-mbig-obj -DCMAKE_EXPORT_COMPILE_COMMANDS=1 -DLLVM_ENABLE_ASSERTIONS=On -DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra" -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=ON -DLLVM_BUILD_EXAMPLES=ON -DLLVM_INCLUDE_EXAMPLES=ON -std=c++20 -march=native),

Bgfx cmake could be found here
https://github.com/hexlord/bgfx.cmake

If I recall one has to use Genie to build bgfx without cmake, I do it the cmake way

___

>BTW, I understand the issue that adding D3D_COMPILER_VERSION == 47 fixes it. The headers are imported from wine. So, I would discuss with them first.

Not sure I follow your thought, please tell me how I could help you here

**Comment by Biswa96** [comment]
Patch sent https://www.winehq.org/pipermail/wine-devel/2020-September/174516.html

**Comment by Hexlord** [comment]
Oh, cool, good catch. So including d3d11shader.h before d3dcompiler.h could lead to incorrect versioning, nice man

**Comment by Biswa96** [comment]
Patch is accepted in wine https://source.winehq.org/git/wine.git/commit/23c6dd55b8c983ec88cada0a6d6c75ee9cd93976. I am not sure if this fixes the issues here.

It will take time to import it mingw-w64-headers and crt packages. If you can't wait then import the patch to mingw-w64 repository and use it to create headers and crt packages.

**Comment by Biswa96** [comment]
Apology for late reply. This issue may be fixed a month ago.

**Comment by RobertBColton** [comment]
Yeah I can confirm it seems to resolved for now for the same project. With the CPPFLAGS override removed, and a clean rebuild, our Direct3D11 is continuing to work fine. The built-in shaders for the engine are compiling and everything. 👍
